### PR TITLE
gl/renderer: Add more color format support for get_surface_data

### PR DESCRIPTION
**Pull request by Zangetsu38**
- Refactor code move if to switch
- Fix regression on T@g with add color format request.
- Add color format request by Home arcade Ingame get fix it.
- Add color format request by gravity rush logo/menu.

# Result: 
image

### CODE REVIEW

File: vita3k/renderer/src/gl/renderer.cpp [hunk 1]
```diff
@@ -482,19 +482,29 @@ void get_surface_data(GLContext &context, size_t width, size_t height, size_t st
...
-        assert(false);
+        break;
+    case SCE_GXM_COLOR_FORMAT_U8U8U8_BGR:
+        glReadPixels(0, 0, static_cast<GLsizei>(width), static_cast<GLsizei>(height), GL_RGB, GL_UNSIGNED_BYTE, pixels);
+        break;
+    case SCE_GXM_COLOR_FORMAT_U8_R:
+        glReadPixels(0, 0, static_cast<GLsizei>(width), static_cast<GLsizei>(height), GL_RED, GL_UNSIGNED_BYTE, pixels);
+        break;
+    default:
+        LOG_ERROR("Color format not implemented: {}, report this to developer", format);
```
Comment (sunho) [inline vita3k/renderer/src/gl/renderer.cpp:505]: We don't need reports for this. We can just complete every color format if we do it wisely.

File: vita3k/renderer/src/gl/renderer.cpp [hunk 2]
```diff
@@ -482,19 +482,29 @@ void get_surface_data(GLContext &context, size_t width, size_t height, size_t st
...
             uint8_t *pixel = reinterpret_cast<uint8_t *>(&pixels[i]);
             std::swap(pixel[0], pixel[3]);
             std::swap(pixel[1], pixel[2]);
         }
-    } else {
-        // TODO wise color conversino implementation
-        // maybe we can use PBO?
-        assert(false);
+        break;
+    case SCE_GXM_COLOR_FORMAT_U8U8U8_BGR:
```
Comment (sunho) [inline vita3k/renderer/src/gl/renderer.cpp:498]: I saw that this is working, but it's really weird. (we should swap some colors in theory) We need to investigate this further.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 1 above)
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:505]: It's maybe that they ignore a part when its using three components. If the stride_in_pixels is the same as in 4 components, this is the right way. We can just swap colors here assuming color format were rendering is bgr. (Thats why using just gl_rgb was working, as it doesnt need to be swapped and it should read first 3 pixel in every stride)

File: vita3k/renderer/src/gl/renderer.cpp [hunk 3]
```diff
@@ -482,19 +482,29 @@ void get_surface_data(GLContext &context, size_t width, size_t height, size_t st
...
+    switch (format) {
+    case SCE_GXM_COLOR_FORMAT_U8U8U8U8_ABGR:
+    case SCE_GXM_COLOR_FORMAT_U8U8U8U8_ARGB:
         glReadPixels(0, 0, static_cast<GLsizei>(width), static_cast<GLsizei>(height), GL_RGBA, GL_UNSIGNED_BYTE, pixels);
-    } else if (format == SCE_GXM_COLOR_FORMAT_U8U8U8U8_RGBA) {
+        break;
+    case SCE_GXM_COLOR_FORMAT_U8U8U8U8_RGBA:
         glReadPixels(0, 0, static_cast<GLsizei>(width), static_cast<GLsizei>(height), GL_RGBA, GL_UNSIGNED_BYTE, pixels);
-        for (int i = 0; i < width * height; ++i) {
+        for (int i = 0; i < width * height; i++) {
```
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:492]: I have all time used i++ for for like that, using ++i have get me crash few time

File: vita3k/renderer/src/gl/renderer.cpp [hunk 4]
```diff
@@ -482,19 +482,29 @@ void get_surface_data(GLContext &context, size_t width, size_t height, size_t st
     }
 
     glPixelStorei(GL_PACK_ROW_LENGTH, static_cast<GLint>(stride_in_pixels));
-    if (format == SCE_GXM_COLOR_FORMAT_U8U8U8U8_ABGR) {
+    switch (format) {
+    case SCE_GXM_COLOR_FORMAT_U8U8U8U8_ABGR:
+    case SCE_GXM_COLOR_FORMAT_U8U8U8U8_ARGB:
```
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Not sur, if works like that with good color wqnt said is correct

File: vita3k/renderer/src/gl/renderer.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:505]: No, i no like add all without any test, so i add it obe by one if game request
For can see what works in réal time

File: vita3k/renderer/src/gl/renderer.cpp (hunk 2 above)
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:498]: No really strange, abgr and argb works for rgba

Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:498]: Im going to do it.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:492]: ++i is i++ without creating temporary value. It's impossible for it to cause crash.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Then somethings going weird in shader recompiler. If we dont swap here we have to remove swap in rgba also.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 1 above)
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:505]: Im going to do it

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: like i thinks,n swap B and R get broken color, (std::swap(pixel[1], pixel[3]);)

File: vita3k/renderer/src/gl/renderer.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:505]: my case, and just with see here for T@g and Home Arcade, need test case by case for all new color ,format request, for don't write wrong code color

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Dont swap in agbr. Swap in argb. Not swapping is simply wrong in our assumption.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:492]: if you said ^^, i have juèst get this crash all time wanted using ++i and tutorial for this for explain like that, so for that i have changed

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: i have swap in ARGB value, and that no works

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: image
swap this = totaly broken

[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Then we're doing it wrong for rgba also. We don't have to swap everywhere and fix shader compiler.

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: i keep it here, only get works correct render without swap

File: vita3k/renderer/src/gl/renderer.cpp (hunk 2 above)
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:498]: Then also remove swap in rgba. Otherwise it makes no sense.

Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:498]: yes i have no understund this swap, but you have write it for what game ?

File: vita3k/renderer/src/gl/renderer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp]: Can you check that your game use the color surface as texture or display. If it's used as display that can be why swapping broke the graphics.

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp]: ah noç, for this, i speek for totaly other code, in profil manager exemple ^^ if using ++i on for i get crash ^^ but here ,i have revert this

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Can you check that your game use the color surface as texture or display. If it's used as display that can be why swapping broke the graphics.

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: how i can check that ?

File: vita3k/renderer/src/gl/renderer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp]: Can you really tell me this specific change caues crash? This specific only ++i. Really?

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp]: no here is good no crash, i have said for other code, so maybe diferent situation

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: You can also check whether color surface data address and texture data address is same. Game can copy data to another buffer. If this happens, you might have to reverse the game.

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: sorry, no understund what you want said

File: vita3k/renderer/src/gl/renderer.cpp (hunk 1 above)
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:505]: Adding each code that works for specific game is workaround. We need to really understand what's going on.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: We're unpacking the texture color format other than abgr. If we don't match the color format and game use color surface as texture. It's going to be broken.

File: vita3k/renderer/src/gl/renderer.cpp (hunk 1 above)
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:505]: i see, but for no, for both, only works like that ^^

File: vita3k/renderer/src/gl/renderer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: really no understund where you waznt check in if you can explain me more on discord please

[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Then game might use color surface for displaying that is other than abgr. We need to also handle this. Or there's something unknown in shader that specify color format.

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: I have really no idea, i just know here that works only with using this, for other, only you can understund more :/

[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/renderer/src/gl/renderer.cpp:487]: Yes I'll explain to you. We have to look into this really deeply. This is closely related to output of shader.

[the remark this answers is not in the document]
Comment (Zangetsu38) [reply vita3k/renderer/src/gl/renderer.cpp:487]: i see

**Pull request merged**